File: scripts/topology_common.py
# SPDX-License-Identifier: Apache-2.0
import json
import os
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SUITE = json.loads((ROOT / 'conformance/vectors/topologies/matrix-0.2.json').read_text(encoding='utf-8'))
RECORDS = json.loads((ROOT / 'conformance/vectors/evaluation/servers-0.1.json').read_text(encoding='utf-8'))
CREDENTIAL = 'synthetic-fixture-credential'


def environment():
    return {**({'SystemRoot':os.environ['SystemRoot']} if 'SystemRoot' in os.environ else {}),'PSP_FIXTURE_CREDENTIAL':CREDENTIAL}


def resource(covenants, capabilities=None):
    return {'classes':[],'covenants':covenants,'capabilities':capabilities or [],'checks':{},'parameters':{},'context':{}}


def approval(name, capabilities=None):
    return {'name':name,'revision':'1','readOnly':name != 'export','complete':True,
        'sources':[{'id':'host-fixture-registry','capabilities':capabilities or []}],
        'inputSchema':RECORDS['inputSchema'],'outputSchema':RECORDS['outputSchema']}


def case_by_id(case_id):
    return next(c for c in SUITE['cases'] if c['id'] == case_id)


def close_approval():
    return {**approval('fixture_close'),'readOnly':False,'inputSchema':{'type':'object','properties':{},'additionalProperties':False},
        'outputSchema':{'type':'object','properties':{'closed':{'type':'boolean'}},'required':['closed'],'additionalProperties':False}}


class EventLog:
    def __init__(self, path, correlation):
        self.file = open(path,'x',encoding='utf-8',newline='\n')
        self.correlation, self.sequence = correlation, 0
    def write(self, kind, record_id=None, code=None):
        if self.sequence >= 128: raise RuntimeError('EVENT_LIMIT')
        self.sequence += 1
        self.file.write(json.dumps({'sequence':self.sequence,'correlation':self.correlation,'kind':kind,'recordId':record_id,'code':code})+'\n')
        self.file.flush()
    def close(self): self.file.close()


def read_events(path, correlation, complete=False):
    path = Path(path)
    text = path.read_text(encoding='utf-8') if path.exists() else ''
    if complete and (not text or not text.endswith('\n')): raise RuntimeError('INCOMPLETE_EVENT_LOG')
    events = [json.loads(line) for line in text.split('\n')[:-1]]
    if any(e['sequence'] != i+1 or e['correlation'] != correlation for i,e in enumerate(events)):
        raise RuntimeError('INVALID_EVENT_LOG')
    return [{k:e[k] for k in ('kind','recordId','code')} for e in events]

File: scripts/topology-common.mjs
// SPDX-License-Identifier: Apache-2.0
import {readFileSync,existsSync,openSync,writeSync,closeSync} from 'node:fs';
export const suite=JSON.parse(readFileSync(new URL('../conformance/vectors/topologies/matrix-0.2.json',import.meta.url),'utf8'));
export const records=JSON.parse(readFileSync(new URL('../conformance/vectors/evaluation/servers-0.1.json',import.meta.url),'utf8'));
export const credential='synthetic-fixture-credential';
export const environment=()=>({... (process.env.SystemRoot?{SystemRoot:process.env.SystemRoot}:{}),PSP_FIXTURE_CREDENTIAL:credential});
export const resource=(covenants,capabilities=[])=>({classes:[],covenants,capabilities,checks:{},parameters:{},context:{}});
export const approval=(name,capabilities=[])=>({name,revision:'1',readOnly:name!=='export',complete:true,
  sources:[{id:'host-fixture-registry',capabilities}],inputSchema:records.inputSchema,outputSchema:records.outputSchema});
export const closeApproval=()=>({...approval('fixture_close'),readOnly:false,inputSchema:{type:'object',properties:{},additionalProperties:false},
  outputSchema:{type:'object',properties:{closed:{type:'boolean'}},required:['closed'],additionalProperties:false}});
export function caseById(id) {
  const c=suite.cases.find(c=>c.id===id);
  if(!c)throw Error('UNKNOWN_CASE');
  return c;
}
export function eventLog(path,correlation) {
  const fd=openSync(path,'wx');let sequence=0;
  return {write(kind,recordId=null,code=null){
    if(sequence>=128)throw Error('EVENT_LIMIT');
    writeSync(fd,JSON.stringify({sequence:++sequence,correlation,kind,recordId,code})+'\n');
  },close(){closeSync(fd);}};
}
export function readEvents(path,correlation,complete=false) {
  const text=existsSync(path)?readFileSync(path,'utf8'):'';
  if(complete&&(!text||!text.endsWith('\n')))throw Error('INCOMPLETE_EVENT_LOG');
  const events=text.split('\n').slice(0,-1).map(JSON.parse);
  if(events.some((e,i)=>e.sequence!==i+1||e.correlation!==correlation))throw Error('INVALID_EVENT_LOG');
  return events.map(({kind,recordId,code})=>({kind,recordId,code}));
}
